beacon/light/canonical.go

```diff
@@ -0,0 +1,134 @@
+
+package light
+
+import (
+	"encoding/binary"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common/lru"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// canonicalStore stores instances of the given type in a database and caches
+// them in memory, associated with a continuous range of period numbers.
+// Note: canonicalStore is not thread safe and it is the caller's responsibility
+// to avoid concurrent access.
+type canonicalStore[T any] struct {
+	keyPrefix []byte
+	periods   Range
+	cache     *lru.Cache[uint64, T]
+	encode    func(T) ([]byte, error)
+	decode    func([]byte) (T, error)
+}
+
+// newCanonicalStore creates a new canonicalStore and loads all keys associated
+// with the keyPrefix in order to determine the ranges available in the database.
+func newCanonicalStore[T any](db ethdb.KeyValueStore, keyPrefix []byte,
+	encode func(T) ([]byte, error), decode func([]byte) (T, error)) *canonicalStore[T] {
+	cs := &canonicalStore[T]{
+		keyPrefix: keyPrefix,
+		encode:    encode,
+		decode:    decode,
+		cache:     lru.NewCache[uint64, T](100),
+	}
+	var (
+		iter  = db.NewIterator(keyPrefix, nil)
+		kl    = len(keyPrefix)
+		first = true
+	)
+	for iter.Next() {
+		if len(iter.Key()) != kl+8 {
+			log.Warn("Invalid key length in the canonical chain database", "key", fmt.Sprintf("%#x", iter.Key()))
+			continue
+		}
+		period := binary.BigEndian.Uint64(iter.Key()[kl : kl+8])
+		if first {
+			cs.periods.Start = period
+		} else if cs.periods.End != period {
+			log.Warn("Gap in the canonical chain database")
+			break // continuity guaranteed
+		}
+		first = false
+		cs.periods.End = period + 1
+	}
+	iter.Release()
+	return cs
+}
+
+// databaseKey returns the database key belonging to the given period.
+func (cs *canonicalStore[T]) databaseKey(period uint64) []byte {
+	var (
+		kl  = len(cs.keyPrefix)
+		key = make([]byte, kl+8)
+	)
+	copy(key[:kl], cs.keyPrefix)
+	binary.BigEndian.PutUint64(key[kl:], period)
+	return key
+}
+
+// add adds the given item to the database. It also ensures that the range remains
+// continuous. Can be used either with a batch or database backend.
+func (cs *canonicalStore[T]) add(backend ethdb.KeyValueWriter, period uint64, value T) error {
+	if !cs.periods.CanExpand(period) {
+		return fmt.Errorf("period expansion is not allowed, first: %d, next: %d, period: %d", cs.periods.Start, cs.periods.End, period)
+	}
+	enc, err := cs.encode(value)
+	if err != nil {
+		return err
+	}
+	if err := backend.Put(cs.databaseKey(period), enc); err != nil {
+		return err
+	}
+	cs.cache.Add(period, value)
+	cs.periods.Expand(period)
+	return nil
+}
+
+// deleteFrom removes items starting from the given period.
+func (cs *canonicalStore[T]) deleteFrom(batch ethdb.Batch, fromPeriod uint64) (deleted Range) {
+	keepRange, deleteRange := cs.periods.Split(fromPeriod)
+	deleteRange.Each(func(period uint64) {
+		batch.Delete(cs.databaseKey(period))
+		cs.cache.Remove(period)
+	})
```

Comment:
```suggestion
func (cs *canonicalStore[T]) deleteFrom(db ethdb.KeyValueWriter, fromPeriod uint64) (deleted Range) {
	keepRange, deleteRange := cs.periods.Split(fromPeriod)
	deleteRange.Each(func(period uint64) {
		db.Delete(cs.databaseKey(period))
		cs.cache.Remove(period)
	})
```
Let's use minimal interfaces

Reply:
done